TEST_APPS/device/nfcapp/nfctestshim.cpp

```diff
@@ -0,0 +1,314 @@
+    } else {
+        cmd_ready(CMDLINE_RETCODE_SUCCESS);
+    }
+}
+
+void NFCTestShim::on_get_rf_protocols()
+{
+#if MBED_CONF_NFCEEPROM
+    cmd_printf("EEPROM cannot get protocol()\r\n");
+    set_last_nfc_error(NFC_ERR_UNSUPPORTED);
+    cmd_ready(CMDLINE_RETCODE_INVALID_PARAMETERS);
+
+#else
+    nfc_rf_protocols_bitmask_t protocols = ((NFCProcessController*)this)->get_rf_protocols();
+    protocols.target_t1t = true;
+    protocols.target_t2t = true;
+    protocols.target_t3t = true;
+    protocols.target_t5t = true;
+    protocols.target_nfc_dep = true;
+    static char strSupported[7*6+1] = "";
+    if (protocols.target_t1t) {
+        strcat(strSupported, "t1t,");
+    }
+    if (protocols.target_t2t) {
+        strcat(strSupported, "t2t,");
+    }
+    if (protocols.target_t3t) {
+        strcat(strSupported, "t3t,");
+    }
+    if (protocols.target_iso_dep) {
+        strcat(strSupported, "isodep,");
+    }
+    if (protocols.target_nfc_dep) {
+        strcat(strSupported, "nfcdep,");
+    }
+    if (protocols.target_t5t) {
+        strcat(strSupported, "t5t,");
+    }
+    if (strlen(strSupported)) {
+        strSupported[strlen(strSupported)-1] = '\0'; // strip trailing comma
+    }
+    cmd_printf("{{protocols=%s}}", strSupported);
+    set_last_nfc_error(NFC_OK);
+    cmd_ready(CMDLINE_RETCODE_SUCCESS);
+#endif
+}
+
+
+void NFCTestShim::on_configure_rf_protocols(nfc_rf_protocols_bitmask_t protocols)
+{
+#if MBED_CONF_NFCEEPROM
+    cmd_printf("EEPROM cannot set protocol()\r\n");
+    set_last_nfc_error(NFC_ERR_UNSUPPORTED);
+    cmd_ready(CMDLINE_RETCODE_INVALID_PARAMETERS);
+#else
+
+    nfc_err_t err = ((NFCProcessController*)this)->set_rf_protocols( protocols );
+    set_last_nfc_error(err );
+    if (NFC_OK != err) {
+        cmd_ready(CMDLINE_RETCODE_FAIL);
+    } else {
+        cmd_ready(CMDLINE_RETCODE_SUCCESS);
+    }
+#endif
+}
+
+// RETURNS: ICETEA error code asynchronously NFC error is set
+// {{bytes=XX XX XX XX.. }} are returned
+void NFCTestShim::on_read_nfceeprom() {
+#if MBED_CONF_NFCEEPROM
+    ((NFCProcessEEPROM*)this)->queue_read_call();
+    cmd_printf("NFCTestShim::read_nfceeprom() exit\r\n");
+
+#else
+    // returns last message "written", since we cannot read
+    print_ndef_message(_ndef_write_buffer, _ndef_write_buffer_used);
+    set_last_nfc_error(NFC_OK);
+
+    cmd_printf("NFCTestShim::read_nfceeprom()\r\n");
+    cmd_ready (CMDLINE_RETCODE_SUCCESS);
+#endif
+}
+
+void NFCTestShim::on_erase() {
+
+#if MBED_CONF_NFCEEPROM
+	((NFCProcessEEPROM*)this)->queue_erase_call();
+
+#else
+	cmd_printf("erase %d bytes, last msg\r\n", (sizeof(_ndef_write_buffer)/sizeof(uint8_t)));
+	_ndef_write_buffer_used = 0;
+    memset(_ndef_write_buffer, 0, sizeof(_ndef_write_buffer)/ sizeof(uint8_t));
+	set_last_nfc_error(NFC_OK); // effectively a no-op
+    cmd_ready(CMDLINE_RETCODE_SUCCESS);
+#endif
+}
+
+// populate buffer with really long message
+void NFCTestShim::on_write_long(int length, char *data)
```

Comment:
please document that data buffer memory ownership is transferred to this function where it's released